packages/lodestar/src/chain/chain.ts

```diff
@@ -328,35 +331,51 @@ export class BeaconChain implements IBeaconChain {
    * Seeds the fork choice with an anchor state.
    * This state is set as the finalized state.
    */
-  private async initForkChoice(anchorState: TreeBacked<BeaconState>): Promise<void> {
-    let blockRoot;
-    let justifiedCheckpoint;
-    let finalizedCheckpoint;
+  private async initForkChoice(
+    anchorState: TreeBacked<BeaconState>
+  ): Promise<{forkChoice: ForkChoice; checkpoint: Checkpoint}> {
     if (anchorState.latestBlockHeader.slot === GENESIS_SLOT) {
       const block = getEmptyBlock();
       block.stateRoot = this.config.types.BeaconState.hashTreeRoot(anchorState);
-      blockRoot = this.config.types.BeaconBlock.hashTreeRoot(block);
+      const blockRoot = this.config.types.BeaconBlock.hashTreeRoot(block);
       const blockCheckpoint = {
         root: blockRoot,
         epoch: computeEpochAtSlot(this.config, anchorState.slot),
       };
-      justifiedCheckpoint = blockCheckpoint;
-      finalizedCheckpoint = blockCheckpoint;
+      const store = new ForkChoiceStore({
+        emitter: this.emitter,
+        currentSlot: this.clock.currentSlot,
+        justifiedCheckpoint: blockCheckpoint,
+        finalizedCheckpoint: blockCheckpoint,
+      });
+      return {
+        forkChoice: ForkChoice.fromGenesis(this.config, store, block),
+        checkpoint: blockCheckpoint,
+      };
     } else {
       const blockHeader = this.config.types.BeaconBlockHeader.clone(anchorState.latestBlockHeader);
-      blockHeader.stateRoot = this.config.types.BeaconState.hashTreeRoot(anchorState);
-      blockRoot = this.config.types.BeaconBlockHeader.hashTreeRoot(blockHeader);
-      justifiedCheckpoint = anchorState.currentJustifiedCheckpoint;
-      finalizedCheckpoint = anchorState.finalizedCheckpoint;
+      if (this.config.types.Root.equals(blockHeader.stateRoot, ZERO_HASH)) {
+        blockHeader.stateRoot = anchorState.hashTreeRoot();
+      }
+      const finalizedCheckpoint = {
+        root: this.config.types.BeaconBlockHeader.hashTreeRoot(blockHeader),
+        epoch: computeEpochAtSlot(this.config, anchorState.slot),
```

Comment:
I'm afraid this is `computeEpochAtSlot(this.config, anchorState.slot) + n` where `n` is unknown if 1st slot of finalized epoch is a skipped slot, it'll use "last known blockRoot" in that case.

I used to face this issue in my weak subjectivity state PR, most peers say goodbye to us due to incorrect finalized checkpoint.

Reply:
ignore the above comment, I think it should work bc the new ArchiveState uses CheckpointStateCache instead of the regular state cache so we're fine but correct me if I'm wrong.

Reply:
The state archival task has been updated so that we only _store_ valid checkpoint states (state where state.slot % SLOTS_PER_EPOCH === 0) to db. Now the epoch will always be correct, if we use those checkpoint states.